File: src/main/python/net/coderextreme/personal/bubble.py
import x3dpsail as x3d
X3D0 = x3d.X3D()
X3D0.setProfile("Immersive")
X3D0.setVersion("3.3")
head1 = x3d.head()
meta2 = x3d.meta()
meta2.setName("title")
meta2.setContent("bubble.x3d")

head1.addMeta(meta2)
meta3 = x3d.meta()
meta3.setName("creator")
meta3.setContent("John Carlson")

head1.addMeta(meta3)
meta4 = x3d.meta()
meta4.setName("description")
meta4.setContent("Tour around a prismatic sphere")

head1.addMeta(meta4)
meta5 = x3d.meta()
meta5.setName("generator")
meta5.setContent("X3D-Edit, https://savage.nps.edu/X3D-Edit")

head1.addMeta(meta5)
meta6 = x3d.meta()
meta6.setName("identifier")
meta6.setContent("https://coderextreme.net/X3DJSONLD/src/main/data/bubble.x3d")

head1.addMeta(meta6)

X3D0.setHead(head1)
Scene7 = x3d.Scene()
NavigationInfo8 = x3d.NavigationInfo()
NavigationInfo8.setType(["EXAMINE"])

Scene7.addChildren(NavigationInfo8)
Viewpoint9 = x3d.Viewpoint()
Viewpoint9.setPosition([0,0,4])
Viewpoint9.setOrientation([1,0,0,0])
Viewpoint9.setDescription("Bubble in action")

Scene7.addChildren(Viewpoint9)
ProtoDeclare10 = x3d.ProtoDeclare()
ProtoDeclare10.setName("Bubble")
ProtoBody11 = x3d.ProtoBody()
Transform12 = x3d.Transform()
Transform12.setDEF("transform")
Shape13 = x3d.Shape()
Sphere14 = x3d.Sphere()
Sphere14.setRadius(0.25)

Shape13.setGeometry(Sphere14)
Appearance15 = x3d.Appearance()
Material16 = x3d.Material()
Material16.setDiffuseColor([1,0,0])
Material16.setTransparency(0.2)

Appearance15.setMaterial(Material16)

Shape13.setAppearance(Appearance15)

Transform12.addChild(Shape13)
Script17 = x3d.Script()
Script17.setDEF("bounce")
field18 = x3d.field()
field18.setName("scale")
field18.setAccessType("inputOutput")
field18.setType("SFVec3f")
field18.setValue("1 1 1")

Script17.addField(field18)
field19 = x3d.field()
field19.setName("translation")
field19.setAccessType("inputOutput")
field19.setType("SFVec3f")
field19.setValue("0 0 0")

Script17.addField(field19)
field20 = x3d.field()
field20.setName("velocity")
field20.setAccessType("inputOutput")
field20.setType("SFVec3f")
field20.setValue("0 0 0")

Script17.addField(field20)
field21 = x3d.field()
field21.setName("scalvel")
field21.setAccessType("inputOutput")
field21.setType("SFVec3f")
field21.setValue("0 0 0")

Script17.addField(field21)
field22 = x3d.field()
field22.setName("set_fraction")
field22.setAccessType("inputOnly")
field22.setType("SFFloat")

Script17.addField(field22)

Script17.setSourceCode('''ecmascript:\n"+
"function initialize() {\n"+
"    velocity = new SFVec3f(Math.random() * 0.25 - 0.125, Math.random() * 0.25 - 0.125, Math.random() * 0.25 - 0.125);\n"+
"\n"+
"    scalvel = new SFVec3f(Math.random() * 0.4, Math.random() * 0.4, Math.random() * 0.4);\n"+
"}\n"+
"\n"+
"function set_fraction(value) {\n"+
"	translation = new SFVec3f(\n"+
"		translation.x + velocity.x,\n"+
"		translation.y + velocity.y,\n"+
"		translation.z + velocity.z);\n"+
"	scale = new SFVec3f(\n"+
"		scale.x + scalvel.x,\n"+
"		scale.y + scalvel.y,\n"+
"		scale.z + scalvel.z);\n"+
"        // if you get to far away or too big, explode\n"+
"        if ( Math.abs(translation.x) > 256) {\n"+
"		translation.x = 0;\n"+
"		initialize();\n"+
"	}\n"+
"        if ( Math.abs(translation.y) > 256) {\n"+
"		translation.y = 0;\n"+
"		initialize();\n"+
"	}\n"+
"        if ( Math.abs(translation.z) > 256) {\n"+
"		translation.z = 0;\n"+
"		initialize();\n"+
"	}\n"+
"	if (Math.abs(scale.x) > 20) {\n"+
"		scale.x = scale.x/2;\n"+
"		translation.x = 0;\n"+
"		initialize();\n"+
"	}\n"+
"	if (Math.abs(scale.y) > 20) {\n"+
"		scale.y = scale.y/2;\n"+
"		translation.y = 0;\n"+
"		initialize();\n"+
"	}\n"+
"	if (Math.abs(scale.z) > 20) {\n"+
"		scale.z = scale.z/2;\n"+
"		translation.z = 0;\n"+
"		initialize();\n"+
"	}\n"+
"}''')

Transform12.addChildren(Script17)
TimeSensor23 = x3d.TimeSensor()
TimeSensor23.setDEF("bubbleClock")
TimeSensor23.setCycleInterval(10)
TimeSensor23.setLoop(True)

Transform12.addChildren(TimeSensor23)
ROUTE24 = x3d.ROUTE()
ROUTE24.setFromNode("bounce")
ROUTE24.setFromField("translation_changed")
ROUTE24.setToNode("transform")
ROUTE24.setToField("set_translation")

Transform12.addChildren(ROUTE24)
ROUTE25 = x3d.ROUTE()
ROUTE25.setFromNode("bounce")
ROUTE25.setFromField("scale_changed")
ROUTE25.setToNode("transform")
ROUTE25.setToField("set_scale")

Transform12.addChildren(ROUTE25)
ROUTE26 = x3d.ROUTE()
ROUTE26.setFromNode("bubbleClock")
ROUTE26.setFromField("fraction_changed")
ROUTE26.setToNode("bounce")
ROUTE26.setToField("set_fraction")

Transform12.addChildren(ROUTE26)

ProtoBody11.addChildren(Transform12)

ProtoDeclare10.setProtoBody(ProtoBody11)

Scene7.addChildren(ProtoDeclare10)
ProtoInstance27 = x3d.ProtoInstance()
ProtoInstance27.setName("Bubble")
ProtoInstance27.setDEF("bubbleA")

Scene7.addChildren(ProtoInstance27)

X3D0.setScene(Scene7)
X3D0.toFileX3D("../personal/bubble.new.python.x3d")
X3D0.toFileJSON("../personal/bubble.new.python.json")

File: src/main/cplusplus/net/coderextreme/personal/bubble.cpp
#define False false
#define True true
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter06/IndexedFaceSetExample/IndexedFaceSetExample/pch.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter06/IndexedFaceSetExample/IndexedFaceSetExample/framework.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter06/IndexedFaceSetExample/include/glut.h"
#include "C:/x3d-code/www.web3d.org/x3d/languages/cpp/Examples_X3DForWebAuthors/Chapter06/IndexedFaceSetExample/include/X3DLib.h"
int main(int argc, char ** argv) {
X3D& X3D0 =  X3D();
X3D0.setProfile(CString("Immersive"));
X3D0.setVersion(CString("3.3"));
head& head1 =  head();
meta& meta2 =  meta();
meta2.setName(CString("title"));
meta2.setContent(CString("bubble.x3d"));
head1.addMeta(&meta2);

meta& meta3 =  meta();
meta3.setName(CString("creator"));
meta3.setContent(CString("John Carlson"));
head1.addMeta(&meta3);

meta& meta4 =  meta();
meta4.setName(CString("description"));
meta4.setContent(CString("Tour around a prismatic sphere"));
head1.addMeta(&meta4);

meta& meta5 =  meta();
meta5.setName(CString("generator"));
meta5.setContent(CString("X3D-Edit, https://savage.nps.edu/X3D-Edit"));
head1.addMeta(&meta5);

meta& meta6 =  meta();
meta6.setName(CString("identifier"));
meta6.setContent(CString("https://coderextreme.net/X3DJSONLD/src/main/personal/bubble.x3d"));
head1.addMeta(&meta6);

X3D0.setHead(&head1);

Scene& Scene7 =  Scene();
NavigationInfo& NavigationInfo8 =  NavigationInfo();
NavigationInfo8.setType(new CString[]{CString("EXAMINE")}, 1);
Scene7.addChild(&NavigationInfo8);

Viewpoint& Viewpoint9 =  Viewpoint();
Viewpoint9.setPosition(new float[]{0,0,4});
Viewpoint9.setOrientation(new float[]{1,0,0,0});
Viewpoint9.setDescription(CString("Bubble in action"));
Scene7.addChild(&Viewpoint9);

ProtoDeclare& ProtoDeclare10 =  ProtoDeclare();
ProtoDeclare10.setName(CString("Bubble"));
ProtoBody& ProtoBody11 =  ProtoBody();
Transform& Transform12 =  Transform();
Transform12.setDEF(CString("transform"));
Shape& Shape13 =  Shape();
Sphere& Sphere14 =  Sphere();
Sphere14.setRadius(0.25);
Shape13.setGeometry(&Sphere14);

Appearance& Appearance15 =  Appearance();
Material& Material16 =  Material();
Material16.setDiffuseColor(new float[]{1,0,0});
Material16.setTransparency(0.2);
Appearance15.addChild(&Material16);

Shape13.addChild(&Appearance15);

Transform12.addChild(&Shape13);

Script& Script17 =  Script();
Script17.setDEF(CString("bounce"));
field& field18 =  field();
field18.setName(CString("scale"));
field18.setAccessType(CString("inputOutput"));
field18.setType(CString("SFVec3f"));
field18.setValue(CString("1 1 1"));
Script17.addChild(&field18);

field& field19 =  field();
field19.setName(CString("translation"));
field19.setAccessType(CString("inputOutput"));
field19.setType(CString("SFVec3f"));
field19.setValue(CString("0 0 0"));
Script17.addChild(&field19);

field& field20 =  field();
field20.setName(CString("velocity"));
field20.setAccessType(CString("inputOutput"));
field20.setType(CString("SFVec3f"));
field20.setValue(CString("0 0 0"));
Script17.addChild(&field20);

field& field21 =  field();
field21.setName(CString("scalvel"));
field21.setAccessType(CString("inputOutput"));
field21.setType(CString("SFVec3f"));
field21.setValue(CString("0 0 0"));
Script17.addChild(&field21);

field& field22 =  field();
field22.setName(CString("set_fraction"));
field22.setAccessType(CString("inputOnly"));
field22.setType(CString("SFFloat"));
Script17.addChild(&field22);


Script17.setSourceCode(CString("ecmascript:")+
_T("function initialize() {")+
_T("    velocity = new SFVec3f(Math.random() * 0.25 - 0.125, Math.random() * 0.25 - 0.125, Math.random() * 0.25 - 0.125);")+
_T("    scalvel = new SFVec3f(Math.random() * 0.4, Math.random() * 0.4, Math.random() * 0.4);")+
_T("}")+
_T("function set_fraction(value) {")+
_T("	translation = new SFVec3f(")+
_T("		translation.x + velocity.x,")+
_T("		translation.y + velocity.y,")+
_T("		translation.z + velocity.z);")+
_T("	scale = new SFVec3f(")+
_T("		scale.x + scalvel.x,")+
_T("		scale.y + scalvel.y,")+
_T("		scale.z + scalvel.z);")+
_T("        // if you get to far away or too big, explode")+
_T("        if ( Math.abs(translation.x) > 256) {")+
_T("		translation.x = 0;")+
_T("		initialize();")+
_T("	}")+
_T("        if ( Math.abs(translation.y) > 256) {")+
_T("		translation.y = 0;")+
_T("		initialize();")+
_T("	}")+
_T("        if ( Math.abs(translation.z) > 256) {")+
_T("		translation.z = 0;")+
_T("		initialize();")+
_T("	}")+
_T("	if (Math.abs(scale.x) > 20) {")+
_T("		scale.x = scale.x/2;")+
_T("		translation.x = 0;")+
_T("		initialize();")+
_T("	}")+
_T("	if (Math.abs(scale.y) > 20) {")+
_T("		scale.y = scale.y/2;")+
_T("		translation.y = 0;")+
_T("		initialize();")+
_T("	}")+
_T("	if (Math.abs(scale.z) > 20) {")+
_T("		scale.z = scale.z/2;")+
_T("		translation.z = 0;")+
_T("		initialize();")+
_T("	}")+
_T("}"));
Transform12.addChild(&Script17);

TimeSensor& TimeSensor23 =  TimeSensor();
TimeSensor23.setDEF(CString("bubbleClock"));
TimeSensor23.setCycleInterval(10);
TimeSensor23.setLoop(True);
Transform12.addChild(&TimeSensor23);

ROUTE& ROUTE24 =  ROUTE();
ROUTE24.setFromNode(CString("bounce"));
ROUTE24.setFromField(CString("translation_changed"));
ROUTE24.setToNode(CString("transform"));
ROUTE24.setToField(CString("set_translation"));
Transform12.addChild(&ROUTE24);

ROUTE& ROUTE25 =  ROUTE();
ROUTE25.setFromNode(CString("bounce"));
ROUTE25.setFromField(CString("scale_changed"));
ROUTE25.setToNode(CString("transform"));
ROUTE25.setToField(CString("set_scale"));
Transform12.addChild(&ROUTE25);

ROUTE& ROUTE26 =  ROUTE();
ROUTE26.setFromNode(CString("bubbleClock"));
ROUTE26.setFromField(CString("fraction_changed"));
ROUTE26.setToNode(CString("bounce"));
ROUTE26.setToField(CString("set_fraction"));
Transform12.addChild(&ROUTE26);

ProtoBody11.addChild(&Transform12);

ProtoDeclare10.addChild(&ProtoBody11);

Scene7.addChild(&ProtoDeclare10);

ProtoInstance& ProtoInstance27 =  ProtoInstance();
ProtoInstance27.setName(CString("Bubble"));
ProtoInstance27.setDEF(CString("bubbleA"));
Scene7.addChild(&ProtoInstance27);

X3D0.setScene(&Scene7);

}
